File: xlang/gate.go
// Go gate — reads the SAME neutral contracts.json and ports the validator.
// Handles the two Go-specific frictions the design flagged:
//   * JSON numbers decode to float64 → the "int" token checks integrality (f == trunc(f)).
//   * missing vs zero → validate map[string]any (not a struct), so absent keys are detectable.
//   conform | produce <route> | consume <route>
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"math"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

type doc struct {
	Contracts map[string]map[string]any `json:"contracts"`
}

var contracts map[string]map[string]any

func jsonish(v any) string { b, _ := json.Marshal(v); return string(b) }

func constEq(tok string, v any) bool {
	if tok == "true" {
		b, ok := v.(bool)
		return ok && b
	}
	if tok == "false" {
		b, ok := v.(bool)
		return ok && !b
	}
	if n, err := strconv.Atoi(tok); err == nil {
		f, ok := v.(float64)
		return ok && f == float64(n)
	}
	s, ok := v.(string)
	return ok && s == tok
}

func leafOk(tok string, v any) bool {
	if strings.HasPrefix(tok, "?") {
		return v == nil || leafOk(tok[1:], v)
	}
	if strings.HasPrefix(tok, "const:") {
		return constEq(tok[6:], v)
	}
	if strings.HasPrefix(tok, "enum:") {
		s, ok := v.(string)
		if !ok {
			return false
		}
		for _, e := range strings.Split(tok[5:], "|") {
			if s == e {
				return true
			}
		}
		return false
	}
	switch tok {
	case "str":
		_, ok := v.(string)
		return ok
	case "int":
		f, ok := v.(float64)
		return ok && f == math.Trunc(f) // JSON int arrives as float64
	case "num":
		_, ok := v.(float64)
		return ok
	case "bool":
		_, ok := v.(bool)
		return ok
	case "obj":
		_, ok := v.(map[string]any)
		return ok
	case "list":
		_, ok := v.([]any)
		return ok
	case "any":
		return true
	}
	return false
}

func check(schema any, value any, where string) error {
	switch sch := schema.(type) {
	case map[string]any:
		if oneOf, ok := sch["oneOf"]; ok {
			var errs []string
			for i, alt := range oneOf.([]any) {
				if err := check(alt, value, fmt.Sprintf("%s|oneOf[%d]", where, i)); err == nil {
					return nil
				} else {
					errs = append(errs, err.Error())
				}
			}
			return fmt.Errorf("%s: matched none of oneOf -> %v", where, errs)
		}
		m, ok := value.(map[string]any)
		if !ok {
			return fmt.Errorf("%s: expected object", where)
		}
		for k, spec := range sch {
			vv, present := m[k]
			if !present {
				if s, ok := spec.(string); ok && strings.HasPrefix(s, "?") {
					continue
				}
				return fmt.Errorf("%s: missing required key '%s'", where, k)
			}
			if err := check(spec, vv, where+"."+k); err != nil {
				return err
			}
		}
		return nil
	case string:
		if !leafOk(sch, value) {
			return fmt.Errorf("%s: %s does not satisfy '%s'", where, jsonish(value), sch)
		}
		return nil
	}
	return fmt.Errorf("%s: unknown schema", where)
}

func examples(c map[string]any) []any {
	if ex, ok := c["examples"].([]any); ok {
		return ex
	}
	return nil
}

func okExample(route string) map[string]any {
	for _, ex := range examples(contracts[route]) {
		e := ex.(map[string]any)
		if r, ok := e["result"].(map[string]any); ok && r["ok"] == true {
			return r
		}
	}
	panic(route + ": no golden ok example")
}

func conform() error {
	for route, c := range contracts {
		effect, _ := c["effect"].(string)
		if effect != "query" && effect != "command" {
			return fmt.Errorf("%s: bad effect", route)
		}
		if strings.Contains(route, "/query/") != (effect == "query") {
			return fmt.Errorf("%s: effect %s contradicts URI verb", route, effect)
		}
		rev, _ := c["reversible"].(bool)
		inv, _ := c["inverseRoute"].(string)
		if rev {
			if _, ok := contracts[inv]; !ok {
				return fmt.Errorf("%s: inverseRoute %s not declared", route, inv)
			}
		}
		for i, ex := range examples(c) {
			e := ex.(map[string]any)
			pl, _ := e["payload"].(map[string]any)
			if pl == nil {
				pl = map[string]any{}
			}
			if err := check(c["inp"], pl, fmt.Sprintf("%s ex[%d].payload", route, i)); err != nil {
				return err
			}
			res, _ := e["result"].(map[string]any)
			if res != nil && res["ok"] == true {
				if err := check(c["out"], res, fmt.Sprintf("%s ex[%d].result", route, i)); err != nil {
					return err
				}
			}
		}
		if rev {
			invInp := contracts[inv]["inp"]
			for i, ex := range examples(c) {
				e := ex.(map[string]any)
				res, _ := e["result"].(map[string]any)
				args := map[string]any{}
				if res != nil {
					if iv, ok := res["inverse"].(map[string]any); ok {
						if a, ok := iv["args"].(map[string]any); ok {
							args = a
						}
					}
				}
				if err := check(invInp, args, fmt.Sprintf("%s ex[%d].inverse.args -> %s input", route, i, inv)); err != nil {
					return err
				}
			}
		}
	}
	return nil
}

func main() {
	here, _ := filepath.Abs(filepath.Dir(os.Args[0]))
	// os.Args[0] is the temp build path under `go run`; read contracts.json next to the source via env.
	path := os.Getenv("XLANG_DIR")
	if path == "" {
		path = here
	}
	raw, err := os.ReadFile(filepath.Join(path, "contracts.json"))
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(2)
	}
	var d doc
	if err := json.Unmarshal(raw, &d); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(2)
	}
	contracts = d.Contracts

	cmd := os.Args[1]
	switch cmd {
	case "conform":
		if err := conform(); err != nil {
			fmt.Fprintln(os.Stderr, "GO  conform FAIL:", err)
			os.Exit(1)
		}
		fmt.Fprintf(os.Stderr, "GO  conform OK — %d contracts\n", len(contracts))
	case "produce":
		b, _ := json.Marshal(okExample(os.Args[2]))
		os.Stdout.Write(b)
	case "consume":
		route := os.Args[2]
		raw, _ := io.ReadAll(os.Stdin)
		var env any
		json.Unmarshal(raw, &env)
		if err := check(contracts[route]["out"], env, "out"); err != nil {
			fmt.Printf(`{"ok":false,"lang":"go","route":%q,"problem":%q}`, route, err.Error())
			os.Exit(1)
		}
		fmt.Printf(`{"ok":true,"lang":"go","route":%q}`, route)
	default:
		fmt.Fprintln(os.Stderr, "unknown cmd", cmd)
		os.Exit(2)
	}
}

File: xlang/gate.mjs
#!/usr/bin/env node
// JS gate — reads the SAME neutral contracts.json and ports the kernel validator 1:1.
// No shared object, no privileged language: the contract file is the only glue.
//   conform                 — validate every contract's golden examples
//   produce <route>         — print the golden ok envelope as JSON
//   consume <route>         — read an envelope on stdin, validate it against <route>.out
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const HERE = path.dirname(fileURLToPath(import.meta.url));
const DOC = JSON.parse(fs.readFileSync(path.join(HERE, 'contracts.json'), 'utf8'));
const CONTRACTS = DOC.contracts;

function parseConst(t) {
  if (t === 'true') return true;
  if (t === 'false') return false;
  if (/^-?\d+$/.test(t)) return parseInt(t, 10);
  return t;
}
function leafOk(tok, v) {
  if (tok.startsWith('?')) return v === null || v === undefined || leafOk(tok.slice(1), v);
  if (tok.startsWith('const:')) return v === parseConst(tok.slice(6));
  if (tok.startsWith('enum:')) return tok.slice(5).split('|').includes(v);
  switch (tok) {
    case 'str': return typeof v === 'string';
    case 'int': return Number.isInteger(v);
    case 'num': return typeof v === 'number';
    case 'bool': return typeof v === 'boolean';
    case 'obj': return v !== null && typeof v === 'object' && !Array.isArray(v);
    case 'list': return Array.isArray(v);
    case 'any': return true;
    default: return false;
  }
}
function check(schema, value, where) {
  const isObj = (x) => x !== null && typeof x === 'object' && !Array.isArray(x);
  if (isObj(schema)) {
    if ('oneOf' in schema) {
      const errs = [];
      for (let i = 0; i < schema.oneOf.length; i++) {
        try { check(schema.oneOf[i], value, `${where}|oneOf[${i}]`); return; }
        catch (e) { errs.push(e.message); }
      }
      throw new Error(`${where}: matched none of oneOf -> ${JSON.stringify(errs)}`);
    }
    if (!isObj(value)) throw new Error(`${where}: expected object, got ${Array.isArray(value) ? 'list' : typeof value}`);
    for (const [k, spec] of Object.entries(schema)) {
      if (!(k in value)) {
        if (typeof spec === 'string' && spec.startsWith('?')) continue;
        throw new Error(`${where}: missing required key '${k}'`);
      }
      check(spec, value[k], `${where}.${k}`);
    }
    return;
  }
  if (!leafOk(schema, value)) throw new Error(`${where}: ${JSON.stringify(value)} does not satisfy '${schema}'`);
}

function conform() {
  for (const [route, c] of Object.entries(CONTRACTS)) {
    if (!['query', 'command'].includes(c.effect)) throw new Error(`${route}: bad effect`);
    if ((route.includes('/query/')) !== (c.effect === 'query'))
      throw new Error(`${route}: effect ${c.effect} contradicts URI verb`);
    if (c.reversible && !(c.inverseRoute in CONTRACTS))
      throw new Error(`${route}: inverseRoute ${c.inverseRoute} not declared`);
    (c.examples || []).forEach((ex, i) => {
      check(c.inp, ex.payload || {}, `${route} ex[${i}].payload`);
      if ((ex.result || {}).ok) check(c.out, ex.result, `${route} ex[${i}].result`);
    });
    if (c.reversible) {
      const inv = CONTRACTS[c.inverseRoute];
      (c.examples || []).forEach((ex, i) => {
        const args = ((ex.result || {}).inverse || {}).args || {};
        check(inv.inp, args, `${route} ex[${i}].inverse.args -> ${c.inverseRoute} input`);
      });
    }
  }
}

function okExample(route) {
  for (const ex of CONTRACTS[route].examples || []) if ((ex.result || {}).ok) return ex.result;
  throw new Error(`${route}: no golden ok example`);
}

function readStdin() {
  return JSON.parse(fs.readFileSync(0, 'utf8'));
}

const [cmd, route] = process.argv.slice(2);
if (cmd === 'conform') {
  conform();
  process.stderr.write(`JS  conform OK — ${Object.keys(CONTRACTS).length} contracts\n`);
} else if (cmd === 'produce') {
  process.stdout.write(JSON.stringify(okExample(route)));
} else if (cmd === 'consume') {
  const env = readStdin();
  try {
    check(CONTRACTS[route].out, env, 'out');
    process.stdout.write(JSON.stringify({ ok: true, lang: 'js', route }));
  } catch (e) {
    process.stdout.write(JSON.stringify({ ok: false, lang: 'js', route, problem: e.message }));
    process.exit(1);
  }
} else {
  throw new Error(`unknown cmd ${cmd}`);
}
